File: training/src/kingdoms_sim/movement.cpp
#include "kingdoms_sim/movement.hpp"

namespace kingdoms {

namespace {

void ChaseTarget(Npc& npc, float speed, float dt, Vec2 target_pos, std::mt19937& rng) {
  const Vec2 aim = NearbyPoint(target_pos, Distance(npc.pos, target_pos), rng);
  const float dx = Sign(aim.x - npc.pos.x);
  const float dy = Sign(aim.y - npc.pos.y);
  npc.pos.x += dx * speed * dt;
  npc.pos.y += dy * speed * dt;
}

void RunFromTarget(Npc& npc, float speed, float dt, Vec2 target_pos, std::mt19937& rng) {
  const Vec2 aim = NearbyPoint(target_pos, Distance(npc.pos, target_pos), rng);
  const float dx = Sign(npc.pos.x - aim.x);
  const float dy = Sign(npc.pos.y - aim.y);
  npc.pos.x += dx * speed * dt;
  npc.pos.y += dy * speed * dt;
}

void OvershootTarget(Npc& npc, float speed, float dt, Vec2 target_pos, std::mt19937& rng) {
  const Vec2 aim = NearbyPoint(target_pos, 1.0f, rng);
  const float dx = Sign(aim.x - npc.pos.x);
  const float dy = Sign(aim.y - npc.pos.y);
  npc.pos.x += dx * speed * dt;
  npc.pos.y += dy * speed * dt;
}

void WanderStep(Npc& npc, float speed, float dt, std::mt19937& rng) {
  if (Distance(npc.movement_point, npc.pos) < 1.0f) {
    npc.movement_point = NearbyPoint(npc.origin, 4.0f, rng);
  }
  const float dx = Sign(npc.movement_point.x - npc.pos.x);
  const float dy = Sign(npc.movement_point.y - npc.pos.y);
  npc.pos.x += dx * speed * dt;
  npc.pos.y += dy * speed * dt;
}

}  // namespace

bool IsHovering(const Npc& npc, const AssetDatabase& assets, Vec2 target_pos) {
  const float dist = Distance(npc.pos, target_pos);
  const float range = assets.Npc(npc.type_id).range;
  const float too_close = range / 2.0f - 1.0f;
  const float too_far = range / 2.0f + 1.0f;
  return dist < too_far && dist > too_close;
}

void MoveNpc(Npc& npc, const AssetDatabase& assets, float dt_seconds, Vec2 player_pos,
             std::mt19937& rng) {
  if (!npc.InCombat()) npc.movement = MovementMode::Wander;

  const float speed = assets.Npc(npc.type_id).modes[npc.mode].speed;

  switch (npc.movement) {
    case MovementMode::Wander:
      WanderStep(npc, speed, dt_seconds, rng);
      break;
    case MovementMode::Chase:
      ChaseTarget(npc, speed, dt_seconds, player_pos, rng);
      break;
    case MovementMode::Run:
      RunFromTarget(npc, speed, dt_seconds, player_pos, rng);
      break;
    case MovementMode::Overshoot:
      OvershootTarget(npc, speed, dt_seconds, player_pos, rng);
      break;
    case MovementMode::Hover: {
      const float dist = Distance(npc.pos, player_pos);
      const float range = assets.Npc(npc.type_id).range;
      if (dist > range / 2.0f + 1.0f) {
        ChaseTarget(npc, speed, dt_seconds, player_pos, rng);
      } else if (dist < range / 2.0f - 1.0f) {
        RunFromTarget(npc, speed, dt_seconds, player_pos, rng);
      }
      break;
    }
    case MovementMode::Turret:
      break;  // stationary; Look()-only in the original, no gameplay effect
  }
}

}  // namespace kingdoms

File: server/engine/movement.go
package engine

import "kingdoms/engine/assets"

func (n *npc) move(seconds float32, data assets.NPC, target *character) {
	if !n.hasTarget || target == nil {
		n.movement = "wander"
		n.wander(seconds, data)
		return
	}

	switch n.movement {
	case "chase":
		n.chase(seconds, data, target.position)
	case "run":
		n.run(seconds, data, target.position)
	case "overshoot":
		n.overshoot(seconds, data, target.position)
	case "hover":
		n.hover(seconds, data, target.position)
	case "turret":
		n.looking = true
	default:
		n.wander(seconds, data)
	}
}

func (n *npc) chase(seconds float32, data assets.NPC, target Position) {
	speed := n.speed(data)
	target = nearby(target, distance(n.position, target))
	n.position.X += direction(target.X-n.position.X) * speed * seconds
	n.position.Y += direction(target.Y-n.position.Y) * speed * seconds
}

func (n *npc) run(seconds float32, data assets.NPC, target Position) {
	speed := n.speed(data)
	target = nearby(target, distance(n.position, target))
	n.position.X += direction(n.position.X-target.X) * speed * seconds
	n.position.Y += direction(n.position.Y-target.Y) * speed * seconds
}

func (n *npc) hover(seconds float32, data assets.NPC, target Position) {
	currentDistance := distance(n.position, target)
	minimum := data.Range/2 - 1
	maximum := data.Range/2 + 1

	switch {
	case currentDistance > maximum:
		n.chase(seconds, data, target)
		n.looking = false
	case currentDistance < minimum:
		n.run(seconds, data, target)
		n.looking = false
	default:
		n.looking = true
	}
}

func (n *npc) hovering(data assets.NPC, target Position) bool {
	currentDistance := distance(n.position, target)
	return currentDistance > data.Range/2-1 && currentDistance < data.Range/2+1
}

func (n *npc) overshoot(seconds float32, data assets.NPC, target Position) {
	speed := n.speed(data)
	target = nearby(target, 1)
	n.position.X += direction(target.X-n.position.X) * speed * seconds
	n.position.Y += direction(target.Y-n.position.Y) * speed * seconds
}

func (n *npc) wander(seconds float32, data assets.NPC) {
	if !n.hasDestination || distance(n.position, n.destination) < 1 {
		n.destination = nearby(n.origin, 4)
		n.hasDestination = true
	}

	speed := n.speed(data)
	n.position.X += direction(n.destination.X-n.position.X) * speed * seconds
	n.position.Y += direction(n.destination.Y-n.position.Y) * speed * seconds
}

func (n *npc) speed(data assets.NPC) float32 {
	if int(n.mode) >= len(data.Modes) {
		return 0
	}
	return data.Modes[n.mode].Speed
}
